test/new-e2e/pkg/test/suite.go

```diff
@@ -0,0 +1,273 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package test
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"reflect"
+	"testing"
+	"time"
+
+	"github.com/DataDog/datadog-agent/test/new-e2e/pkg/runner"
+	"github.com/DataDog/datadog-agent/test/new-e2e/pkg/runner/parameters"
+	"github.com/DataDog/test-infra-definitions/common/utils"
+	"github.com/DataDog/test-infra-definitions/components"
+
+	"github.com/stretchr/testify/suite"
+)
+
+const (
+	importKey = "import"
+
+	createTimeout          = 60 * time.Minute
+	deleteTimeout          = 30 * time.Minute
+	provisionerGracePeriod = 2 * time.Second
+)
+
+// Suite is a generic inteface used internally, only implemented by BaseSuite
+type Suite[Env any] interface {
+	suite.TestingSuite
+
+	init(params []Option, self Suite[Env])
+
+	UpdateEnv(map[string]Provisioner)
+	Env() *Env
+}
+
+var _ Suite[any] = &BaseSuite[any]{}
+
+type BaseSuite[Env any] struct {
+	suite.Suite
+
+	env    *Env
+	params suiteParams
+
+	currentProvisioners map[string]Provisioner
+	targetProvisioners  map[string]Provisioner
+
+	firstFailTest string
+}
+
+// Custom methods
+func (bs *BaseSuite[Env]) Env() *Env {
+	bs.reconcileEnv()
+	return bs.env
+}
+
+func (bs *BaseSuite[Env]) UpdateEnv(newProvisioners map[string]Provisioner) {
+	bs.targetProvisioners = newProvisioners
```

Comment:
<!-- Nitpicks are small, trivial, but necessary changes. Distinguishing nitpick comments significantly helps direct the reader's attention to comments requiring more involvement. -->
**🥜 nitpick**
For non-lazy mode we can reconcile the env right here at UpdateEnv